Mathlib/LinearAlgebra/Matrix/Swap.lean

```diff
@@ -0,0 +1,164 @@
+/-
+Copyright (c) 2024 Judith Ludwig, Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Judith Ludwig, Christian Merten
+-/
+import Mathlib.LinearAlgebra.Matrix.GeneralLinearGroup.Defs
+
+/-!
+# Swap matrices
+
+A swap matrix indexed by `i` and `j` is the matrix that, when multiplying another matrix
+on the left (resp. on the right), swaps the `i`-th row with the `j`-th row
+(resp. the `i`-th column with the `j`-th column).
+
+Swap matrices are a special case of *elementary matrices*. For transvections see
+`Mathlib.LinearAlgebra.Matrix.Transvection`.
+-/
+
+namespace Matrix
+
+section Def
+
+variable {R n : Type*} [Ring R] [DecidableEq n]
+
+variable (R) in
+/-- The swap matrix `swap R i j` is the identity matrix with the
+`i`-th and `j`-th rows modified such that multiplying by it on the
+left (resp. right) corresponds to swapping the `i`-th and `j`-th row (resp. column). -/
+def swap (i j : n) : Matrix n n R :=
+  1 - stdBasisMatrix i i 1
+    - stdBasisMatrix j j 1
+    + stdBasisMatrix i j 1
+    + stdBasisMatrix j i 1
```

Comment:
I think at the very least we should define this in terms of `(Equiv.swap _ _).toPEquiv.toMatrix`, as then it doesn't need `Ring` or even `Neg`, and will work on matrices of `Nat`.

I'm still not sure whether we want `swap` at all, but I think it will be easier to tell once it's written in terms of the existing API.

Reply:
Indeed, using `PEquiv.toMatrix` significantly improves things. I generalized some of the lemmas to that setting and now just specialize.